### ISSUE TITLE
no-param-reassign inconsistency with redeclared variables

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#11942]
**Tell us about your environment**

* **ESLint Version:** 6.0.1
* **Node Version:** 10.16.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  parserOptions: {
    ecmaVersion: 6,
  },
  rules: {
    "no-param-reassign": "error"
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function foo(a, b) {
    var a = 1;
    var b;
    b = 1;
}
```

```bash
eslint index.js
```

**What did you expect to happen?**

I guess 2 errors, `var` does not declare a new variable here, it's still the parameter.

Maybe 0 errors.

**What actually happened? Please include the actual, raw output from ESLint.**

1 Error.

```js
function foo(a, b) {
    var a = 1; // no error
    var b;
    b = 1; // error
}
```

**Are you willing to submit a pull request to fix this bug?**

Yes, just to find out first what is the intended behavior.

There is an explicit `!reference.init` check in the code. But, it might be there to prevent false positives with param default values `f(a = 1)`, and is causing false negatives with initialized redeclarations.

**Comment by mdjermanovic** [comment]
The same issue with other "no assign" rules:


```js
/* eslint no-global-assign: "error" */

var Object = 1; // no error
Object = 1; // error
```
```js
/* eslint no-ex-assign: "error" */

try {
} catch (e) {
    var e = 1; // no error
    e = 1; // error
}
```

```js
/* eslint no-func-assign: "error" */

function f() {}

var f = 1; // no error
f = 1; // error
```

`no-const-assign` and `no-class-assign` don't have this issue, because it is already a parsing error.

**Comment by mdjermanovic** [comment]
Side note regarding the ScopeManager.

```js
try {
} catch (e) {      // (1)
    var e = 1;     // (2) 
}
```

`variable` declared in line `(1)` has a `reference` in line `(2)`

`reference.init` is maybe wrongly set to `true`, because the line `(2)` actually represents a declaration of another variable and an assignment to the variable from line `(1)`.

**Comment by platinumazure** [comment]
> The same issue with other "no assign" rules:

We have `no-shadow` for cases where a variable with the same name being declared in a new variable scope.

**Comment by mdjermanovic** [comment]
> We have `no-shadow` for cases where a variable with the same name being declared in a new variable scope.

It's the same scope, so `no-shadow` does not report any of these 4 examples (including the first post). It might report only the `catch` example (upside-down, the `catch` variable is indeed shadowing the `var` variable), but it doesn't work. Anyway, I think that `no-shadow` is not supposed to report the remaining 3 examples.  but `no-redeclare` (included in `eslint:recommended`) would report them.

However, if the `no-redeclare` is `off`, these 4 rules from the examples will allow reassignments.

If this is a feature, i.e. the rules do not treat the assignments as assignments to param/global/catch/function anymore, because the same name is redeclared as `var`, then it might be inconsistent.

For example, this is not reported:

```js
/* eslint no-global-assign: "error" */

var Object = 1; // no error
```

but this is reported:

```js
/* eslint no-global-assign: "error" */

var Object;
Object = 1; // error
```

**Comment by kaicataldo** [comment]
Reopening so we can verify if this is a bug or not

**Comment by platinumazure** [comment]
@mysticatea You're the scope/variables expert these days :smile: Do you have any thoughts on whether this is a bug, or working as intended?
